core/src/main/scala/kafka/log/LogCleanerManager.scala

```diff
@@ -337,6 +384,32 @@ private[log] class LogCleanerManager(val logDirs: Seq[File],
       }
     }
   }
+
+  def markPartitionUncleanable(logDir: String, partition: TopicPartition): Unit = {
+    inLock(lock) {
+      uncleanablePartitions.get(logDir) match {
+        case Some(partitions) =>
+          partitions.add(partition)
+        case None =>
+          uncleanablePartitions.put(logDir, mutable.Set(partition))
+      }
+    }
+  }
+
+  /*
+    Returns an immutable set of the uncleanable partitions for a given log directory
+   */
+  def uncleanablePartitions(logDir: String): Set[TopicPartition] = {
+    var partitions: Set[TopicPartition] = Set()
+    inLock(lock) { partitions ++= uncleanablePartitions.getOrElse(logDir, partitions) }
+    partitions
+  }
+
+  private def isUncleanablePartition(log: Log, topicPartition: TopicPartition): Boolean = {
+    def isUncleanable: Boolean = uncleanablePartitions.get(log.dir.getParent).exists(partitions => partitions.contains(topicPartition))
+
+    if (lock.isLocked) isUncleanable else inLock(lock) { isUncleanable }
```

Comment:
Do we need conditional locking given `lock` is a `ReentrantLock`?

Reply:
Oh right. my bad